### DISCUSSION TITLE
Hooks for "close others" and "close all"?

### DISCUSSION THREAD

**Discussion by hoelzro** [TiddlyWiki/TiddlyWiki5#7166] [Ideas]
Hello!

I noticed that while closing a single tiddler fires a `th-closing-tiddler` hook, the "close others" and "close all" actions don't seem to fire any hooks.  Would it make sense to add hooks for these actions as well?  If so, I wonder whether having distinct `th-close-others` and `th-close-all` hooks would make sense, or just having those actions fire repeated `th-closing-tiddler` hooks (which of course would have implications around the event parameter passed to hook functions and those functions' return values)

**Comment by pmario** [comment]
>Would it make sense to add hooks for these actions as well?

What would you like to do with those hooks? ... 

At the moment, the philosophy is, that hooks only get added, if either the core needs and uses it __or__ if there are some strong arguments from plugin-authors. 

"Because we can" ... wouldn't make much sense, because it makes the core code more complex, without a real immediate benefit. ... 

As you can see at the existing hooks, they have different return values and even the "function signature" is different for some of them. ... So they are probably more "opinionated" as they should be. ... That's the way how the system has "grown" over time. ...

**Comment by hoelzro** [comment]
That's a fair question - here are two examples of plugins I've been thinking this would be useful for:

  * I have an "undo close" plugin I wrote that currently uses `$tw.wiki.addEventListener('change',...)` to monitor `$:/StoryList` to detect closing of tiddlers, and I feel using hooks would be a more elegant/proper solution
  * I'm working on a "breadcrumbs" plugin to display information on why I opened a tiddler, and I'd like to detect closing of tiddlers so that I can clean up metadata associated with a tiddler once it's closed

I can, of course, make use of the `change` event to detect closing of tiddlers, but that feels pretty low-level, especially since `th-closing-tiddler` is already available, and it feels a little strange to me that if someone asks the question "how do I detect when a tiddler is closed?" the answer is "hook into `th-closing-tiddler`, but only if you care about the close button - if you care about close all/close others, you'll need to use another method"

> As you can see at the existing hooks, they have different return values and even the "function signature" is different for some of them. ... So they are probably more "opinionated" as they should be. ... That's the way how the system has "grown" over time. ...

I think you have a really good point here - if hooks were extended to include close others/close all, the interface would have to be chosen carefully.

**Comment by Jermolene** [comment]
The biggest problem with hooks is that they are tied to TiddlyWiki's JavaScript implementation. Our long term roadmap is to migrate the implementation of TiddlyWiki's high level features to wikitext, so that they can be more easily understood and hacked by end users. For example, I've been exploring a wikitext implementation of the navigator widget in [TWPub](https://github.com/TWPUB/TWPUB-Tools/blob/main/plugins/twpub-tools/%24-plugins-immateriel-twpub-tools/globals/widgets/multinavigator.tid). These wikitext implementations also need to be extensible, but the mechanisms will need to be different.

With that in mind, I wonder if we might think about another approach to the OP here. As you mention, the desired functionality could in principle be obtained by monitoring changes in the story list tiddler. So, we could try to make a general purpose low level mechanism that would help us do that.

One way of approaching it might be a mechanism that monitors the results of a filter over time, and invokes action strings as items are added or removed from the results.

These observable filters might be defined by tiddlers with a tag like `$:/tags/ObservableFilter` with a `filter` field containing the text of the filter and the `text` field containing the actions to be performed. There would be an observable filter daemon that would listen to change events from the wiki in order to track observable filters. It would maintain a cache of the results of the filter at the last refresh cycle so that it could figure out which were the additions and removals and dispatch the action strings as required.

**Comment by hoelzro** [comment]
Ok, thanks for the additional context @Jermolene! I'd be content with a wikitext-based solution, as long as I can consistently respond to closing tiddlers, and that "observable filter" idea sounds like it would fit the bill 🙂

**Reply by Jermolene** [reply]
Thanks @hoelzro so as not to monopolise the thread I've made a ticket for the proposal here: #7170

**Comment by AnthonyMuscio** [comment]
Rather than hooks although I see no harm in them I think we should lead with the tiddlywiki way first. 
* I believe any hook is in tiddlywiki can be made available in JavaScript but the first priority is making it available to any user

My earlier proposal needs some developer support (although not much) to introduce actions to any button via tags. As demonstrated here https://all-button-actions.tiddlyhost.com/

If you were to use the above solution it would be a matter of writing a tiddler containing the "additional" actions and tagging it with both `$:/tags/button-actions/close-all` and `$:/tags/button-actions/close-others`

[$:/core/ui/Buttons/close-all](https://all-button-actions.tiddlyhost.com/#%24%3A%2Fcore%2Fui%2FButtons%2Fclose-all)
[$:/core/ui/Buttons/close-others](https://all-button-actions.tiddlyhost.com/#%24%3A%2Fcore%2Fui%2FButtons%2Fclose-others)
